kernel/pci/src/lib.rs

```diff
@@ -530,4 +560,77 @@ impl DerefMut for PciDevice {
 pub enum PciConfigSpaceAccessMechanism {
     MemoryMapped = 0,
     IoPort = 1,
-}
\ No newline at end of file
+}
+
+/// A memory-mapped array of [`MsixVectorEntry`]
+pub struct MsixVectorTable {
+    inner: BorrowedSliceMappedPages<MsixVectorEntry, Mutable>,
+}
+
+impl MsixVectorTable {
+    pub fn new(inner: BorrowedSliceMappedPages<MsixVectorEntry, Mutable>) -> Self {
+        Self {
+            inner,
+        }
+    }
+}
+
+impl Deref for MsixVectorTable {
+    type Target = [MsixVectorEntry];
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl DerefMut for MsixVectorTable {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+/// A single Message Signaled Interrupt entry.
+/// It contains the interrupt number for this vector and the core this interrupt is redirected to.
+#[derive(FromBytes)]
+#[repr(C)]
+pub struct MsixVectorEntry {
+    /// The lower portion of the address for the memory write transaction.
+    /// This part contains the CPU ID which the interrupt will be redirected to.
+    msg_lower_addr:         Volatile<u32>,
+    /// The upper portion of the address for the memory write transaction.
+    msg_upper_addr:         Volatile<u32>,
+    /// The data portion of the msi vector which contains the interrupt number.
+    msg_data:               Volatile<u32>,
+    /// The control portion which contains the interrupt mask bit.
+    vector_control:         Volatile<u32>,
+}
+
+impl MsixVectorEntry {
+    /// Sets interrupt destination & number for this entry
+    pub fn configure(&mut self, cpu_id: CpuId, int_num: InterruptNumber) {
+        // unmask the interrupt
+        self.vector_control.write(MSIX_UNMASK_INT);
+        let lower_addr = self.msg_lower_addr.read();
+
+        // set the core to which this interrupt will be sent
+        let dest_id = (cpu_id.into_u8() as u32) << MSIX_DEST_ID_SHIFT;
```

Comment:
The day CpuId isn't 8-bits anymore, this could cause hard to diagnose bugs as this code expects the cpu id to only occupy 8 bits in this u32 (other bits are used for other purposes). This `into_u8` will loudly fail once this happens, which is better imo.

Reply:
oh god, why on earth is this a thing.... sigh. x86 blows, lol. I had hoped to remove `CpuId::into_u8()`, but now we can't? hmm.

So I'm struggling to actually find documentation on the bit-wise format of this field. Specifically I don't see where the CPU ID gets involved; perhaps @Ramla-I can chime in too since she originally implemented this?

Reply:
The Intel Architecture manual ([Vol 3a, 10.11](https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3a-part-1-manual.pdf)) has the information on the layout of the MSI fields.

The destination ID field refers to to the target processor.

Reply:
Thanks, I see now. It's not necessarily an 8-bit CPU ID, it's just an IOAPIC-defined destination ID. Since we currently only use physical addressing mode to specify an interrupt's destination, it's the same thing as a CPU ID (an APIC ID on x86).